### PR TITLE
[Network] Add Cmdlets for BYOIP Feature

### PR DESCRIPTION
**Pull request by dexiang-ms**
## Description

This PR adds the code for the Powershell Cmdlets for the BYOIP feature and is in accordance with the following review PRs:

**Initial proposal to add Master and CustomIpPrefix:** https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/632
**Update PublicIpPrefix:** https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/670
**Removal of MasterCustomIpPrefix:** https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/696

## Checklist

- [X] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [X] The title of the PR is clear and informative
- [X] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [X] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [X] If applicable, the changes made in the PR have proper test coverage
- [X] For public API changes to cmdlets:
    - [X] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [X] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

File: src/Network/Network/ChangeLog.md [hunk 1]
```diff
@@ -20,6 +20,14 @@
 
 ## Upcoming Release
 
+* Added new cmdlets to support BYOIP functionality (currently in private preview, not available to the public)
+    - `New-AzCustomIpPrefix`
+    - `Get-AzCustomIpPrefix`
+    - `Update-AzCustomIpPrefix`
+    - `Remove-AzCustomIpPrefix`
+* Modified the New-AzPublicIpPrefix cmdlet to support BYOIP functionality
+	- Add optional property "CustomIpPrefix" to allow future customers to associate the prefix with the CustomIpPrefix range that the customer owns.
```
Comment (isra-fel) [inline src/Network/Network/ChangeLog.md:29]: Nit: indentation
```suggestion
    - Add optional property "CustomIpPrefix" to allow future customers to associate the prefix with the CustomIpPrefix range that the customer owns.
```

File: src/Network/Network/Properties/Resources.Designer.cs [hunk 2]
```diff
@@ -267,6 +267,15 @@ internal class Resources {
             }
         }
         
+        /// <summary>
+        ///   Looks up a localized string similar to Cannot have both &apos;Commission&apos; and &apos;Decomission&apos; flags set.
+        /// </summary>
+        internal static string CommissioningStateConflict {
+            get {
+                return ResourceManager.GetString("CommissioningStateConflict", resourceCulture);
```
Comment (isra-fel) [inline src/Network/Network/Properties/Resources.Designer.cs:275]: Did you not check in the change in `Resources.resx`?

File: src/Network/Network/PublicIpPrefix/NewAzurePublicIpPrefixCommand.cs [hunk 3]
```diff
@@ -96,6 +96,12 @@ public class NewAzurePublicIpPrefixCommand : PublicIpPrefixBaseCmdlet
             ValueFromPipelineByPropertyName = true)]
         public string[] Zone { get; set; }
 
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The CustomIpPrefix that this PublicIpPrefix will be associated with")]
+        public PSCustomIpPrefix CustomIpPrefix { get; set; }
```
Comment (isra-fel) [inline src/Network/Network/PublicIpPrefix/NewAzurePublicIpPrefixCommand.cs:103]: Help doc of this cmdlet should be updated.
Would be great to add an example

**Comment by isra-fel** [comment]
BTW I rembmer NRP used to have a seperate branch e.g. `network-july` for all the features. Are you not following that practice any more?

**Comment by dexiang-ms** [comment]
> BTW I rembmer NRP used to have a seperate branch e.g. `network-july` for all the features to be merged to. Are you not following that practice any more?

Oh you might be right, would you prefer that I re-do this PR onto that branch?

File: src/Network/Network/ChangeLog.md (hunk 1 above)
Comment (dexiang-ms) [reply src/Network/Network/ChangeLog.md]: fixed

File: src/Network/Network/Properties/Resources.Designer.cs (hunk 2 above)
Comment (dexiang-ms) [reply src/Network/Network/Properties/Resources.Designer.cs:275]: you're right, fixed

File: src/Network/Network/PublicIpPrefix/NewAzurePublicIpPrefixCommand.cs (hunk 3 above)
Comment (dexiang-ms) [reply src/Network/Network/PublicIpPrefix/NewAzurePublicIpPrefixCommand.cs:103]: updated help doc - I think we want to hold off on the example; this feature is still in private preview so I don't want to make it too obvious for the world to see and use this feature. Furthermore, there's a chance that things change so I kind of want to keep this feature somewhat hidden for the time being. Once we go into public preview/GA, then we will solidify all the documentation.

**Comment by isra-fel** [comment]
> Oh you might be right, would you prefer that I re-do this PR onto that branch?

Cool. Simply click the "Edit" button and choose `network-june` as the base branch.

**Comment by dexiang-ms** [comment]
> > Oh you might be right, would you prefer that I re-do this PR onto that branch?
> 
> Cool. Simply click the "Edit" button and choose `network-june` as the base branch.
> FYI - there are build issues. Please check the pipelines

Do you mean 'network-july'? Or has network-june not been merged yet

**Comment by isra-fel** [comment]
> Also about the build issues, it seems like the build is failing because it cant find references to the SDK files. It seems like the SDK for Network has not been released yet, latest is still the May release (https://www.nuget.org/packages/Microsoft.Azure.Management.Network/). I'm assuming I have to wait until the Network team updates their SDK before I can merge this PR?

I think NRP has a practice of using local SDK package in network branch. @MikhailTryakhov could you point the instructions to @dexiang-ms ? Thanks

Besides, your branch was not created from `Azure:network-june` so changing the base branch causes too many commits in this PR. I suggest create a new one from `Azure:network-june` and cherry-pick _your_ commits there.

**Comment by dexiang-ms** [comment]
> > Also about the build issues, it seems like the build is failing because it cant find references to the SDK files. It seems like the SDK for Network has not been released yet, latest is still the May release (https://www.nuget.org/packages/Microsoft.Azure.Management.Network/). I'm assuming I have to wait until the Network team updates their SDK before I can merge this PR?
> 
> I think NRP has a practice of using local SDK package in network branch. @MikhailTryakhov could you point the instructions to @dexiang-ms ? Thanks
> 
> Besides, your branch was not created from `Azure:network-june` so changing the base branch causes too many commits in this PR. I suggest create a new one from `Azure:network-june` and cherry-pick _your_ commits there. Then submit a new PR

I think the local SDK practice is only for local development, but these merge validation checks are using the published SDK. Waiting on @MikhailTryakhov to update us on the progress of the SDK.

**Comment by dexiang-ms** [comment]
Re-created PR: https://github.com/Azure/azure-powershell/pull/12704/files

**Pull request closed without merging**